hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpConnectionChannelErrorHandler.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.nio.tcp;
+
+import com.hazelcast.instance.OutOfMemoryErrorDispatcher;
+import com.hazelcast.internal.networking.Channel;
+import com.hazelcast.internal.networking.ChannelErrorHandler;
+import com.hazelcast.logging.ILogger;
+
+import java.io.EOFException;
+
+public class TcpIpConnectionChannelErrorHandler implements ChannelErrorHandler {
+
+    private final ILogger logger;
+
+    public TcpIpConnectionChannelErrorHandler(ILogger logger) {
+        this.logger = logger;
+    }
+
+    @Override
+    public void onError(Channel channel, Throwable cause) {
+        if (cause instanceof OutOfMemoryError) {
+            OutOfMemoryErrorDispatcher.onOutOfMemory((OutOfMemoryError) cause);
+        }
+
+        if (channel == null) {
+            // todo: question is if logging is the best solution. If an exception happened without a channel, it is a pretty
+            // big event and perhaps we should shutdown the whole HZ instance.
+            logger.severe(cause);
+        } else {
+            TcpIpConnection connection = (TcpIpConnection) channel.attributeMap().get(TcpIpConnection.class);
+            if (cause instanceof EOFException) {
+                connection.close("Connection closed by the other side", cause);
+            } else {
+                connection.close("Exception in " + getClass().getSimpleName(), cause);
```

Comment:
Good spotting. 

I'll include the threadname (current thread) and fix the channel to give a better 'toString'